aips/aip-x-safe-auth-key-mapping.md

```diff
@@ -0,0 +1,221 @@
+   the original one-to-one mapping assumption.
+
+# Summary
+
+The onchain key rotation address mapping has functional issues which inhibit
+safe mapping of authentication key to originating address for standard accounts.
+I propose resolving these issues with the reference implementation.
+
+# Proposed solution
+
+Assorted checks and extra function logic in [`aptos-core` #14309]
+
+# Alternative solutions
+
+Per @davidiw in a separate chat:
+
+> something where we update the rotation command to take in new and removed keys
+> as well as any additional metadata that might be useful such as parameters for
+> the authkey type. Then we add to the indexer a set of all keys that point to
+> potential addresses.
+
+As I understand, this approach would require breaking changes and would
+introduce offchain indexing as an additional dependency in the authentication
+key mapping paradigm.
+
+My solution, captured in the proposed reference implementation, offers a
+purely onchain solution to existing issues and does not require altering the
+existing design space or introducing an offchain dependency.
+
+# Specification
+
+N/A
+
+# Reference implementations
+
+[`aptos-core` #14309]
+
+# Risks and drawbacks
+
+Enforces a one-to-one mapping of private key to account address in the general
+case of following best practices, which extreme users (wishing to use one
+private key to authenticate all their accounts) may find restrictive.
+
+# Security considerations
+
+Note that the function `account::set_originating_address` proposed in
+[`aptos-core` #14309] must remain a private entry function to prevent unproven
+key rotation attacks.
+
+# Multisig considerations
+
+In a separate chat, @davidiw asked about how the changes in the reference
+implementation will interact with multisig v2. Note that even without the
+changes in this PR, it is already possible for a multisig to have an entry in
+the `OriginatingAddress` table:
+
+1. Rotate account `A` to have a new authentication key, thus generating an entry
+   in the `OriginatingAddress` table.
+2. Convert account `A` to a multisig via
+   `multisig_account::create_with_existing_account_and_revoke_auth_key`, which
+   will set the account's authentication key to `0x0`, but which will *not*
+   mutate the `OriginatingAddress` table, since it makes an inner call to
+   `account::rotate_authentication_key_internal`.
+3. The `OriginatingAddress` table then (incorrectly) reports that a mapping from
+   the authentication key (from before multisig conversion) to the multisig
+   address.
+
+# Timelines
+
+Ideally during next release
+
+# Future potentials
+
+Potentially in a separate update, logic to eradicate the existing multisig v2
+indexing issues mentioned above (which is outside the scope of what the
+reference implementation intends to resolve).
+
+# Verifying changes in reference implementation
+
+As requested by @thepom on 2024-09-17:
```

Comment:
Once again, I would remove such references; the AIP should stand on its own.

Reply:
Addressed in 4e5fb6a